### ISSUE TITLE
Cleaner thread to purge old project files

### ISSUE THREAD

**Issue by mukund-thakur** [azkaban/azkaban#934]
Currently we have only option of purging old execution logs using this parameter execution.logs.retention.ms. We must have a feature to clean older project files also as table project_files keep on increasing rapidly similar to purging old execution logs.

**Comment by ameyamk** [comment]
Hi Mukund - would you be interested in contributing this feature?

**Comment by mukund-thakur** [comment]
Yes , I am interested in contributing this feature.

**Comment by mukund-thakur** [comment]
Here is my design approach:
projects.retention.ms parameter will be accepted from the config similar to execution.logs.retention.ms.
A thread will be started in ProjectManager during web server start and will run daily which will cleanup all the older projects.

**Comment by HappyRay** [comment]
Great. This feature will be useful.

How do you decide which project to delete? 

What does projects.retention.ms represent? What do you think of a max size of the cache config? 

Please see the new config name naming convention at
https://github.com/azkaban/azkaban/blob/master/azkaban-common/src/main/java/azkaban/constants/ServerProperties.java and other files in this directory.

**Comment by mukund-thakur** [comment]
Please find the response inline.

How do you decide which project to delete?
-- We will get the list of projects to be purged based on projects.retention.ms config. Following query will be used "select name from projects where active = 1 and modified_time < (current_timestamp - projects.retention.ms)  order by id limit.

What does projects.retention.ms represent?
-- Represents the retention time in milliseconds. We will delete all the older unmodified projects than this time. Default value of this can be 3 months.

What do you think of a max size of the cache config?
-- Please clarify which cache config are u talking about.

**Comment by HappyRay** [comment]
I see. 

I misunderstood the issue. I thought you were referring to the unbounded growth of the project directory ( cache ) disk space on the executors.  By max size of the cache config I was referring to setting up a max limit on how much disk space the project cache can consume.

===

You are referring to deleting project data in the DB. Right?

Unlike logs, there is an expectation that the DB contains the source of truth for a project definition and unless the project owner deletes the project, that data should be kept. 

I have to check the code to verify when the modified_time is updated. The name suggests that it is updated whenever a new version of a project is uploaded. Is it safe to delete when such a project may still be actively used?

**Comment by mukund-thakur** [comment]
@HappyRay 
I will explain the real motivation behind this feature request. Recently we had shortage of disk in our azkaban mysql machine. After some initial debugging we found that project_files is taking too much of space ( 80 % of total db space). Then we found that azkaban has api's to delete and purge project files. We just wrote a simple program to do the cleanup but external programs are really difficult to maintain. Then i though this should be a feature in azkaban only which anybody can use based on configuration. 

You are right about the modified_time. It is updated when a project is deleted or a newer version of it is created. Also the delete and purge api internally has some check. The cleaner feature will be using the implementation of these api's only of regular deletion.

**Comment by HappyRay** [comment]
It's a known issue with the current design. We are working on a solution that will allow ivy coordinates of jars to be specified for a project instead of including them in the zip. 

There is a configuration that controls how many old versions of projects to keep. How many versions of your old projects do you current keep?

How do you make sure that the deleted project is not going to be used? Are you just deleting the project files or deleting the entire project?

**Comment by mukund-thakur** [comment]
@HappyRay  I don't really understand what does this ivy coordinates mean here. I think project files needs to be present in db as we have multi executors here. 

Currently we keep 3 versions.

The delete and purge api makes sure that projects which is getting deleted is not being used. Also it just deletes the project files , project permissions and project properties not the entire project.

Anyways , this feature will be configurable. If the user doesn't want to delete older projects he can set the values of projects.retention.ms parameter to a very high value.

**Comment by HappyRay** [comment]
Are you referring to azkaban.project.ProjectManager#purgeProject?
This will only delete the older version of projects, right? 

We only keep the active project files. All important project files should be in source control.

**Comment by mukund-thakur** [comment]
Yes, i am talking about azkaban.project.ProjectManager#purgeProject . 
This deletes all the versions of project files not just the oldest version.

**Comment by ameyamk** [comment]
I am worried that this config may inadvertently delete project files. Reason being there is implicit contract on UI when user uploads project files - and this feature will override that contract - and to make it really safe we may need to create alerts etc. #NoOfProjectVersions seems more safe and manageable approach - which already exists today. In our production we just keep 1 old version - perhaps you can try that as well?
